web/src/pill_typeahead.ts

```diff
@@ -0,0 +1,220 @@
+import assert from "minimalistic-assert";
+
+import * as blueslip from "./blueslip";
+import * as people from "./people";
+import type {User} from "./people";
+import * as stream_pill from "./stream_pill";
+import type {StreamSubscription} from "./sub_store";
+import * as typeahead_helper from "./typeahead_helper";
+import type {UserOrMention} from "./typeahead_helper";
+import * as user_group_pill from "./user_group_pill";
+import * as user_groups from "./user_groups";
+import type {UserGroup} from "./user_groups";
+import * as user_pill from "./user_pill";
+
+function person_matcher(query: string, item: User): boolean {
+    if (people.is_known_user(item)) {
+        return typeahead_helper.query_matches_person(query, item);
+    }
+    return false;
+}
+
+function group_matcher(query: string, item: UserGroup): boolean {
+    if (user_groups.is_user_group(item)) {
+        return typeahead_helper.query_matches_name(query, item);
+    }
+    return false;
+}
+
+type StreamRecipient = StreamSubscription & {type: "stream"};
+export type UserGroupRecipient = UserGroup & {type: "user_group"};
+export type UserRecipient = UserOrMention & {type: "user"};
+
+export type TypeaheadItem = StreamRecipient | UserGroupRecipient | UserRecipient;
+
+export function set_up(
+    $input: JQuery<HTMLInputElement>,
```

Comment:
It turns out this is not in fact an `<input>`, it’s `<div class="input" contenteditable="true">`.